File: leetcode/Sqrt(x).java
import java.util.*;

public class Solution {
    public int sqrt(int x) {
        if (x <= 1) return x;
        int a = 0;
        int b = x;
        while (a + 1 < b) {
            int m = (a + b) / 2;
            if (m <= x / m) a = m;
            else b = m;
        }
        return a;
    }
}

File: leetcode/Sqrt(x).cpp
class Solution {
    int find(int beg, int end, int target) {
        if (beg + 1 == end)
            return beg;
        long long mid = ((long long)beg + end) / 2;
        if (mid * mid > target)
            return find(beg, mid, target);
        else
            return find(mid, end, target);
    }
public:
    int sqrt(int x) {
        if (x < 2)
            return x;
        return find(1, x, x);
    }
};
